omnipaxos/src/omni_paxos.rs

```diff
@@ -73,18 +69,77 @@ impl OmniPaxosConfig {
     }
 }
 
-impl Default for OmniPaxosConfig {
+/// Configuration for an `OmniPaxos` cluster.
+/// # Fields
+/// * `configuration_id`: The identifier for the cluster configuration that this Sequence Paxos replica is part of.
+/// * `nodes`: The nodes in the cluster i.e. the `pid`s of the other replicas in the configuration.
+/// * `initial_leader`: The initial leader of the cluster. Could be used in combination with reconfiguration to skip the prepare phase when switching to a new configuration.
+#[derive(Clone, Debug, PartialEq)]
+#[cfg_attr(any(feature = "serde", feature = "toml_config"), derive(Deserialize))]
+#[cfg_attr(feature = "toml_config", serde(default))]
+#[cfg_attr(feature = "serde", derive(Serialize))]
+pub struct ClusterConfig {
+    /// The identifier for the cluster configuration that this Sequence Paxos replica is part of.
+    pub configuration_id: u32,
+    /// The nodes in the cluster i.e. the `pid`s of the other replicas in the configuration.
+    pub nodes: Vec<u64>,
+    /// The initial leader of the cluster. Could be used in combination with reconfiguration to skip the prepare phase when switching to a new configuration.
+    pub initial_leader: Option<Ballot>,
+}
+
+impl ClusterConfig {
+    /// Checks all configurations and returns the local OmniPaxos node if successful.
+    pub fn build_for<T, B>(self, storage: B, server_config: ServerConfig) -> OmniPaxos<T, B>
+    where
+        T: Entry,
+        B: Storage<T>,
+    {
+        let op_config = OmniPaxosConfig {
+            cluster_config: self,
+            server_config,
+        };
+        op_config.build(storage)
+    }
+}
+
+impl Default for ClusterConfig {
+    fn default() -> Self {
+        Self {
+            configuration_id: 1,
+            nodes: Vec::new(),
+            initial_leader: None,
+        }
+    }
+}
```

Comment:
Do we need default? I would maybe prefer to not provide a default implementation since users are supposed to set this. If we need it to derive serde then let's it's better to derive default, where configuration_id = 0

Reply:
The advantage of default is it allows users to more simply configure only the things they care about. Like how you've shown in the book before:
```rust
let omnipaxos_config = OmniPaxosConfig {
    configuration_id,
    pid: my_pid,
    peers: my_peers,
    ..Default::default()
}
```
Also it gives an idea to users about what a reasonable config is. The problem is that there really is no default `pid`, or `nodes`.

Reply:
Yes, that is precisely my point, there are no good default values for pid, peers, and configuration_id. We should therefore not provide a default implementation for it. When they code and get forced to specify these values (rather than just use default to make the compiler happy), they will look at the docs/examples to see how to set them, which is what we want.

Default for other configs make sense when we have many fields and there are good default values that can be set. This is not the case here in my opinion.

Reply:
But as I said, if we really need it for ser/de then keep it but make the default values fail to build.